lib/ansible/module_utils/dimensiondata.py

```diff
@@ -0,0 +1,393 @@
+
+        if not found_networks:
+            raise UnknownNetworkError("Network '%s' could not be found" % name)
+
+        net_id = found_networks[0].id
+
+    return driver.ex_get_network_domain(net_id)
+
+
+def get_vlan(driver, locator, location, network_domain):
+    """
+    Get a VLAN object by its name or id
+    """
+    if is_uuid(locator):
+        vlan_id = locator
+    else:
+        vlans = driver.ex_list_vlans(location=location,
+                                     network_domain=network_domain)
+        found_vlans = filter(lambda x: x.name == locator, vlans)
+
+        if not found_vlans:
+            raise UnknownVLANError("VLAN '%s' could not be found" % locator)
+
+        vlan_id = found_vlans[0].id
+
+    return driver.ex_get_vlan(vlan_id)
+
+
+def get_mcp_version(driver, location):
+    """
+    Get a locations MCP version
+    """
+    # Get location to determine if MCP 1.0 or 2.0
+    location = driver.ex_get_location_by_id(location)
+    if 'MCP 2.0' in location.name:
+        return '2.0'
+    return '1.0'
+
+
+def is_uuid(u, version=4):
+    """
+    Test if valid v4 UUID
+    """
+    try:
+        uuid_obj = UUID(u, version=version)
+        return str(uuid_obj) == u
+    except:
+        return False
+
+
+def expand_ip_block(block):
+    """
+    Expand public IP block to show all addresses
+    """
+    addresses = []
+    ip_r = block.base_ip.split('.')
+    last_quad = int(ip_r[3])
+    address_root = "%s.%s.%s." % (ip_r[0], ip_r[1], ip_r[2])
+    for i in range(int(block.size)):
+        addresses.append(address_root + str(last_quad + i))
+    return addresses
+
+
+def get_public_ip_block(module, driver, network_domain, block_id=False,
+                        base_ip=False):
+    """
+    Get public IP block details
+    """
+    # Block ID given, try to use it.
+    if block_id is not 'False':
+        try:
+            block = driver.ex_get_public_ip_block(block_id)
+        except DimensionDataAPIException as e:
+            # 'UNEXPECTED_ERROR' should be removed once upstream bug is fixed.
+            # Currently any call to ex_get_public_ip_block where the block does
+            # not exist will return UNEXPECTED_ERROR rather than
+            # 'RESOURCE_NOT_FOUND'.
+            if e.code == "RESOURCE_NOT_FOUND" or e.code == 'UNEXPECTED_ERROR':
+                module.exit_json(changed=False, msg="Public IP Block does "
+                                 "not exist")
+            else:
+                module.fail_json(msg="Unexpected error while retrieving "
+                                 "block: %s" % e.code)
+            module.fail_json(msg="Error retreving Public IP Block " +
+                                 "'%s': %s" % (block.id, e.message))
+    # Block ID not given, try to use base_ip.
+    else:
+        blocks = list_public_ip_blocks(module, driver, network_domain)
+        if blocks is not False:
+            block = filter(lambda x: x.base_ip == base_ip, blocks)[0]
```

Comment:
This will fail on python3 for a similar reason as the conditional check noted above.  filter returns a generator.  The generator can't be indexed.

Use a list comprehension instead:
``` python
block = [b for b in blocks if b.base_ip == base_ip]
```